server/py/framework/utils/clients/discovery.py

```diff
@@ -0,0 +1,150 @@
+# Copyright 2024 Iguazio
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import collections
+import re
+import typing
+from dataclasses import dataclass
+
+import mlrun.utils.singleton
+from mlrun import mlconf
+
+
+@dataclass
+class ServiceInstance:
+    name: str
+    url: str
+    method_routes: dict[str, list[re.Pattern]] = None
+    status: str = "UP"  # UP, DOWN, UNKNOWN
+
+
+class Client(
+    metaclass=mlrun.utils.singleton.AbstractSingleton,
+):
+    def __init__(self):
+        self._logger = mlrun.utils.logger.get_child(__name__)
+        self.services = None
+        self.initialize()
+
+    def initialize(self):
+        # We use an ordered dict for control over service matching order.
+        # This is important for services that may have overlapping routes.
+        self.services: dict[str, ServiceInstance] = collections.OrderedDict()
+        if mlconf.services.hydra.services != "*":
+            self.register_service(service_name="alerts")
+        self.register_service(service_name="api")
+        # Chief is last to allow other services to override its routes
+        self.register_service(service_name="api-chief")
```

Comment:
I think it is a bit strange that the `discovery` client "know" each micro service.
Why the  `register_service` isn't call from the micro service itself?
If it is because overlapping routes and for making some order, anyway I don't think it is good idea that we will have overlapping routers, maybe to solve it we should add the service name prefix before the route?

Reply:
This is a dummy discovery client. It's not possible to call the discovery from the services since they are not initialized in all instances. The real way to do this is when the service starts it sends a request to register to a remote KV DB like etcd. But this will not happen in 1.8